### PR TITLE
Improve error message under 401 condition with Azure ACR

### LINKED ISSUE #6902: Bad error message when showing recipes in ACR
### Steps to reproduce (required)
- Login with ACR login
- Publish by running `rad bicep publish`
- Run the following command

```cli
rad recipe show <recipeName> --resource-type <resourceType>
```

### Observed behavior (required)
```
\"<acrName>.azurecr.io/azure/sql:latest\": HEAD \"https://<acrName>.azurecr.io/v2/azure/<recipeName>/manifests/latest\": GET \"https://<acrName>.azurecr.io/oauth2/token?scope=repository%3Aazure%2F<recipeName>%3Apull\u0026service=<acrName>.azurecr.io\": response status code 401: unauthorized: authentication required, visit https://aka.ms/acr/authorization for more information."
```

### Desired behavior (required)
A better error message that explains the problem and tells you how to login to pull. If I were an external customer I wouldn't be sure how to fix this.

[AB#10641](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/10641)

### PR DESCRIPTION
**Pull request by gpltaylor**
# Description

Improve error message returned when publishing to a private ACR

## Type of change

- This pull request is a minor refactor, code cleanup, test improvement, or other maintenance task and doesn't change the functionality of Radius (issue link optional).

Fixes: #6902 #6297

## Auto-generated summary

copilot:all

### CODE REVIEW

File: pkg/cli/cmd/bicep/publish/publish.go
Comment (rynowak) [inline pkg/cli/cmd/bicep/publish/publish.go:171]: Hi @gpltaylor thanks for sending this PR. Apologies for the delay in giving feedback, it's the holiday's here.

I have a few concerns about this approach.

----

First, I'd like to see if we can make the error reporting better for all hosts, not just ACR. ACR is one of the common OCI registries we've seen in use, but in general any OCI registry works.

Secondly, is there a better way to check for authentication errors? Any time we have to do a string comparison on an error message to see what kind of error it is, it's going to be fragile. The project we're relying on (ORAS) can change the error message at any time.

I think the best way to figure this out would be to set a breakpoint, reproduce the problem, and then see what type of error we get back. My guess is that it's going to be one of these: https://github.com/radius-project/radius/issues/6297#issuecomment-1866873796

When you know what type of error is being returned, you can use `errors.Is` to check for it. 


https://github.com/oras-project/oras-go/blob/main/registry/remote/errcode/errors.go#L42

Comment (gpltaylor) [reply pkg/cli/cmd/bicep/publish/publish.go:171]: I agree using string is a poor choice however this error response from ACR has been standard. 
This was a quick fix for ACR and the typical 401 returned. This is why I kept the standard error. 

I will see if I can review the response and see if there is something more strongly typed.

I am still focusing on ACR as this is the problem set. Should I try to make this more generic for other registers?

Comment (rynowak) [reply pkg/cli/cmd/bicep/publish/publish.go:171]: > I agree using string is a poor choice however this error response from ACR has been standard.
This was a quick fix for ACR and the typical 401 returned. This is why I kept the standard error.

> I will see if I can review the response and see if there is something more strongly typed.

If you're able to reproduce this, you should be able to able to handle a more specific error type. That way we can be use it's an authorization problem and not some other kind of problem.

> I am still focusing on ACR as this is the problem set. Should I try to make this more generic for other registers?

We support **any** OCI registry. If you can make generic improvements that would be better. I'm not opposed to printing a nice error for ACR's case as well, but this should get better for any registry.

**Comment by gpltaylor** [comment]
I have added the construct to improve error messaging to a user publishing.
 
There are a number of Container Registers that all act differently and return "stringly" errors. I have included a base for extending this. I have debugged against GCR, ACR and Docker Hub and all return different error messages.

For example google can return the following, or which returns a status 400. oras-go does not provide any way to interprit this message.

```
Failed to publish Bicep file "../../test/myapp/app.bicep" to "gcr.io/my-project/my-new-image:test" Cause: HEAD "https://gcr.io/v2/my-project/my-new-image/manifests/sha256:4629660f49ab61ec5abb96e207cd795c15d95fa0be1647999a9665359f04ad83": GET "https://gcr.io/v2/token?scope=repository%3Amy-project%2Fmy-new-image%3Apull&service=gcr.io": response status code 400: unknown: Service 'containerregistry.googleapis.com' is not enabled for consumer 'project:my-project'..
```

It appears that improving the messaging to the user will require string pattern matching. As this is brittle I have omitted but I have included ACR using a Host match.

However, I have included better messaging for ACR and included more "user friendly" messaging for common errors. 
Using Host and HttpStatus we can extend this quickly in fucture.

I look forward to feedback.

**Comment by gpltaylor** [comment]
polite knock knock @rynowak 
PS Is there a process for PR changes?

File: pkg/cli/cmd/bicep/publish/publish.go
Comment (rynowak) [inline pkg/cli/cmd/bicep/publish/publish.go:174]: This is looking a lot better overall 👍 

This code will still need to handle the case where the error is NOT `*errcode.ErrorResponse`. Otherwise cases like a network failure will be reported as success.

You might also want to refactor the error handling for `httpErr` into it's own function for readability.

It might look something like this:

```go
err = r.publish(ctx)
var httpErr *errcode.ErrorResponse
if errors.As(err, &httpErr) {
  return handleErrorResponse(httpErr)
} else if err != nil {
  return clierrors.MessageWithCause(err, "Failed to publish Bicep file %q to %q.", r.File, r.Target)
}
```

```
func handleErrorResponse(err *errcode.ErrorResponse) error {
... // Your current code goes here.
}
```

File: pkg/cli/cmd/bicep/publish/publish.go
Comment (rynowak) [inline pkg/cli/cmd/bicep/publish/publish.go:197]: I tried this out, and this is what it looks for for ACR:

image

As discussed [here](https://github.com/radius-project/radius/issues/6297#issuecomment-1866873796) I think the error message should clearly explain what's failing. I'd suggest bringing back the `Failed to publish Bicep file %q to %q.` as the first line. 

I think you can also remove `Error: %q`. `clierrors.MessageWithCause` will already nicely format the error you give it.

**Review by rynowak — CHANGES_REQUESTED** [review]
This is much improved and on the right track overall. Left some feedback about what the output looks like.

The general error case also needs to be handled before I can merge. Thanks for your help so far 👍

File: pkg/cli/cmd/bicep/publish/publish.go
Comment (gpltaylor) [reply pkg/cli/cmd/bicep/publish/publish.go]: Changed made as requested. 

> This code will still need to handle the case where the error is NOT *errcode.ErrorResponse.
OMG 💯

File: pkg/cli/cmd/bicep/publish/publish.go
Comment (gpltaylor) [reply pkg/cli/cmd/bicep/publish/publish.go]: > I think you can also remove Error: %q. clierrors.MessageWithCause
Done

> As discussed https://github.com/radius-project/radius/issues/6297#issuecomment-1866873796 I think the error message should clearly explain what's failing

You've said this a few times in comments now however, I believe this is done now. 
If I am missing the point please be blunt :)

Comment (gpltaylor) [reply pkg/cli/cmd/bicep/publish/publish.go]: @rynowak When refactoring code into a function, should I introduce unit testing for that function? Does a test here, add value?

File: pkg/cli/cmd/bicep/publish/publish.go
Comment (rynowak) [inline pkg/cli/cmd/bicep/publish/publish.go:176]:
```suggestion
		message := fmt.Sprintf("Failed to publish Bicep file %q to %q", r.File, r.Target)
```

This should use quotes to surround the user-specified details like the filename and destination.

File: pkg/cli/cmd/bicep/publish/publish.go
Comment (rynowak) [inline pkg/cli/cmd/bicep/publish/publish.go:189]: This is looking much better, here's what it looked like for me when I tried it out.

image

I'd suggest removing the quotes around `message` and `acrInfo` (replace `%q` with `%s`). These are user-facing explanations and it's confusing that they are surrounded with an extra layer of quotes.

**Comment by rynowak** [comment]
> polite knock knock @rynowak PS Is there a process for PR change notification - aka code updated following a first PR? This is not covered here: https://github.com/radius-project/radius/tree/main/docs/contributing/contributing-pull-requests

Apologies, it's the holidays here so I'm a little bit slower, you did the right thing by sending a message.

File: pkg/cli/cmd/bicep/publish/publish.go
Comment (rynowak) [reply pkg/cli/cmd/bicep/publish/publish.go]: > If I am missing the point please be blunt :)

I think this has been addressed now 👍 

> When refactoring code into a function, should I introduce unit testing for that function? Does a test here, add value?

In general yes it's a good thing, the value is that we'll be able to make future changes with confidence.

Give it a try, and let me know if you get stuck or want advice.

**Comment by gpltaylor** [comment]
Happy New Year @rynowak

**Comment by rynowak** [comment]
FYI - since this is your first pull-request to Radius it's asking me to approve actions runs. I've approved these so you can see the build/tests running when you push changes.

The functional tests require approval as well, but there should no be impact on those from your changes. I'll approve that as the last step when this is ready to merge.

File: pkg/cli/cmd/bicep/publish/publish.go
Comment (gpltaylor) [reply pkg/cli/cmd/bicep/publish/publish.go]: I will add some tests and update the PR

File: pkg/cli/clierrors/types.go
Comment (rynowak) [inline pkg/cli/clierrors/types.go:47]: 👍

File: pkg/cli/cmd/bicep/publish/publish_test.go
Comment (rynowak) [inline pkg/cli/cmd/bicep/publish/publish_test.go:330]: Looks like this should be `return e.Error` and you can remove the commented-out code.

File: pkg/cli/cmd/bicep/publish/publish_test.go
Comment (rynowak) [inline pkg/cli/cmd/bicep/publish/publish_test.go:346]: Use [subtests](https://go.dev/blog/subtests) for each of these. Go has a feature for this :)

File: pkg/cli/cmd/bicep/publish/publish_test.go
Comment (rynowak) [inline pkg/cli/cmd/bicep/publish/publish_test.go:377]: Is this meant to say `Failure` instead of `Faure`?

File: pkg/cli/cmd/bicep/publish/publish_test.go
Comment (rynowak) [inline pkg/cli/cmd/bicep/publish/publish_test.go:391]: I'd suggest adding a description or name field here. Your comments like `// Test case 5: Other status code` can just become the test name.

File: pkg/cli/cmd/bicep/publish/publish_test.go
Comment (rynowak) [inline pkg/cli/cmd/bicep/publish/publish_test.go:397]: I think this could probably more readable by just moving all of the data inline. 

```go
// Case
{
  name: "...",
  httpErr: &MockErrorResponse{
		URL:        &url.URL{Host: "myregistry.azurecr.io"},
		StatusCode: http.StatusInternalServerError,
	},
  message: "Failure reason5",
  expectedErr: fmt.Sprintf(...),
}
```

File: pkg/cli/cmd/bicep/publish/publish_test.go
Comment (rynowak) [inline pkg/cli/cmd/bicep/publish/publish_test.go:405]: Is there a reason to use `MockErrorResponse` at all? It looks like we can just use `&errcode.ErrorResponse` directly.

File: pkg/cli/cmd/bicep/publish/publish_test.go
Comment (rynowak) [inline pkg/cli/cmd/bicep/publish/publish_test.go:409]: Use `require.Equal()`

**Review by rynowak — CHANGES_REQUESTED** [review]
Awesome 👍 thanks for adding tests. Left a few more comments.

**Comment by gpltaylor** [comment]
@rynowak latest updates to the unit tests.

Let me know if there is anything else... Thank you so much for the feedback so far!

**Review by rynowak — APPROVED** [review]
Awesome! Thanks @gpltaylor

**Comment by rynowak** [comment]
Merged, thanks @gpltaylor !

**Pull request merged**
